### PR TITLE
Add WebAuthn support to GUI (fixes #8409)

### LINKED ISSUE #8409: WebAuthn support in GUI
I would love to see the GUI support [WebAuthn](https://webauthn.guide/) as an alternative to password authentication. Although the Syncthing GUI perhaps doesn't need the strongest security since it's typically not exposed to the whole internet, WebAuthn can still offer a much more pleasant user experience than passwords for those users that prefer it. Personally I use a password manager protected by a YubiKey, and it's rather annoying to have to take the detour through the password manager when I could just tap the YubiKey instead - especially when accessing the Syncthing GUI remotely on my phone, I wouldn't have to switch back and forth between apps to retrieve a password.

I have a working, fairly polished [implementation](https://github.com/emlun/syncthing/tree/webauthn) ready, though I expect it'll take a bit more work for it to pass review (most notably, I haven't added any tests yet). Shall I open a pull request, or would you like to discuss it some more first?

### PR DESCRIPTION
**Pull request by emlun**
Fixes #8409.

### Remaining things to do

- [x] Rewrite data type changes in protobuf
- [x] Store and validate signature counter
- [ ] ~~Store transports and return in `allowCredentials`~~ Not currently supported by [duo-labs/webauthn](https://github.com/duo-labs/webauthn)
- [x] Remove backwards-incompatible template string syntax in JS
- [x] Remove basic auth prompt
- [x] Proper password auth API endpoint
- [x] WebAuthn failure messages on login page
- [x] Warning on login page when on raw IP address
- [ ] Migrate from deprecated duo-labs/webauthn to successor fork
- [ ] Add tests
- [ ] Open documentation PR

### Purpose

This adds support for [WebAuthn](https://webauthn.guide/) authentication in the GUI. WebAuthn is a W3C standard that provides an API for strong, privacy-conscious public-key authentication, which can enable a both more pleasant and more secure user experience than traditional password authentication. WebAuthn both works with external security keys (typically via USB or NFC) and is built into many modern browsers and operating systems.

For users who already use WebAuthn, this both cuts out a detour through a password manager and improves security by making it possible to not need a password at all.

### Testing

Manual tests performed:

- With no WebAuthn credentials enrolled:
  - If no password set, no authentication is required
  - ~~Basic authorization prompt appears like before~~ Removed!
- With WebAuthn credentials enrolled:
  - With password set, Basic authorization headers still work as before
  - An HTML login page with both WebAuthn and password options is presented
  - Password authentication in HTML form works (no WebAuthn required)
  - WebAuthn authentication in HTML form works (no password required)
  - WebAuthn authentication is available immediately without server restart
- WebAuthn settings/credential management:
  - Warning appears when HTTPS is disabled
  - Warning appears when RP ID does not match page domain
  - Warning appears when WebAuthn origin does not match RP ID
  - Warning appears when changing RP ID
  - New credentials appear immediately in the table in GUI settings
  - Renaming credentials takes effect on saving the GUI settings
  - Credential names fall back to credential ID when not set
  - Deleting credentials takes effect on saving the GUI settings
  - `PUT /rest/config` ignores WebAuthn credentials not already registered (identified by ID)
  - [`excludeCredentials`][exclude] is used to prevent registering the same authenticator twice

I intend to add automated tests as well, but would first like a review of the overall design and how it fits into the existing architecture.

### Screenshots

HTML login form
HTML login form

New GUI settings
New GUI settings

Registering a new credential
Registering a new credential

Renaming a credential
Renaming a credential

Interactive warnings
Interactive warnings

### Documentation

Probably a new section should be added alongside "LDAP Authentication". WebAuthn normally doesn't require much technical knowledge from the user since you only interact with the end-user side, but in the Syncthing case you are both end-user and server administrator. WebAuthn requires some domain settings to line up, so the "RP ID" and "WebAuthn Origin" settings and their impact need to be explained here. I volunteer to write this documentation once I have the green light to finish the feature.

[exclude]: https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-excludecredentials
[transports]: https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-transports
[signCount]: https://w3c.github.io/webauthn/#signcount

### CODE REVIEW

File: lib/config/guiconfiguration.pb.go [hunk 1]
```diff
@@ -23,6 +26,12 @@ var _ = math.Inf
 // proto package needs to be updated.
 const _ = proto.GoGoProtoPackageIsVersion3 // please upgrade the proto package
 
+type WebauthnCredential struct {
+	ID                      string                      `protobuf:"bytes,1,name=id,proto3" json:"id" xml:"id,attr" nodefault:"true"`
+	Nickname                string                      `protobuf:"bytes,2,opt,name=nickname,proto3" json:"nickname" xml:"nickname,attr"`
+	PublicKeyCose           string                      `protobuf:"bytes,3,opt,name=publicKeyCose,proto3" json:"publicKeyCose" xml:"publicKeyCose,attr" restart:"false"`
+}
+
```
[the remark this answers is not in the document]
Comment (emlun) [reply lib/config/guiconfiguration.pb.go:34]: Oh, and I'm not sure whether the `protobuf` tags are needed here, or on the new `GUIConfiguration` fields. At least, it seems I didn't have to make any changes to the `XXX_*` methods in order for WebAuthn to work. Do we need those for this?

**Comment by tomasz1986** [comment]
Translations are handled by Transifex, so you shouldn't add anything manually, except for the English strings, which are optional also (as the script will add them later anyway).

Another question is whether this should be added to the "main" GUI (as the commit currently does) or rather stay in the Advanced Configuration only (as it is the case with LDAP).

**Comment by calmh** [comment]
That's awesome! I'll look into this and answer questions. In the meantime, after just glancing at your description, I think it would be nice to axe the http basic auth prompt completely. That is, if we have a real login page, we should present that always. (Possibly excepting the case where basic auth is present without us having prompted for it, like in a wrapper that adds it automatically.)

File: lib/config/guiconfiguration.pb.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (calmh) [reply lib/config/guiconfiguration.pb.go:34]: These `.pb.go`-files are auto generated. The proper way to add fields is to add them to the `.proto` files under `proto/` at the repo top level, then run `go run build.go proto` (which may need some dependencies, I don't fully remember...).

**Comment by tomasz1986** [comment]
This seems to have poor browser support though. I've tested the code locally, and the GUI doesn't load in IE11 at all. Also, according to https://webauthn.me/browser-support, the minimum Android requirement is version 7+, while Syncthing itself currently supports Android 4.1.

Would it be possible to load the new authentication only when the browser/wrapper supports it?

**Comment by emlun** [comment]
> I think it would be nice to axe the http basic auth prompt completely

Great, I agree! And yes, we can easily keep checking for the presence of the `Authorization` header and accept that if present, but just not return the `WWW-Authenticate` header if not.

And right, I meant to point something out in the diff but forgot to: currently the login form works by sending an XHR with the `Authorization` header, just to get the cookie, and then reloading the page with JavaScript - just because that was easy to get working and didn't need any more changes to the backend. But especially if we're dropping the basic auth prompt, I think it's probably cleaner to add a proper auth endpoint to the backend and have the `<form>` actually POST to that instead.

>the GUI doesn't load in IE11 at all. [...]
> Would it be possible to load the new authentication only when the browser/wrapper supports it?

Hm, I thought [this check](https://github.com/syncthing/syncthing/pull/8417/files#diff-77bf455925e2b856359d0a3f63bd07b75fe2990adf70d4996de0df0a0cfe8b29R1477) would fix that, but honestly I forgot to test it because I don't have a non-compatible browser on hand. Do you get any error in the JavaScript console?

**Comment by tomasz1986** [comment]
> Hm, I thought [this check](https://github.com/syncthing/syncthing/pull/8417/files#diff-77bf455925e2b856359d0a3f63bd07b75fe2990adf70d4996de0df0a0cfe8b29R1477) would fix that, but honestly I forgot to test it because I don't have a non-compatible browser on hand. Do you get any error in the JavaScript console?

Sure, this is the output.

```
SCRIPT1002: Syntax error
webauthn-json.browser-global.js (1,5)
SCRIPT1014: Invalid character
syncthingController.js (1546,28)
Error: [ng:areq] Argument 'SyncthingController' is not a function, got undefined
http://errors.angularjs.org/1.3.20/ng/areq?p0=SyncthingController&p1=not%20a%20function%2C%20got%20undefined
   at assertArg (https://127.0.0.1:8384/vendor/angular/angular.js:1609:5)
   at assertArgFn (https://127.0.0.1:8384/vendor/angular/angular.js:1619:3)
   at Anonymous function (https://127.0.0.1:8384/vendor/angular/angular.js:8512:9)
   at Anonymous function (https://127.0.0.1:8384/vendor/angular/angular.js:7680:13)
   at forEach (https://127.0.0.1:8384/vendor/angular/angular.js:353:11)
   at nodeLinkFn (https://127.0.0.1:8384/vendor/angular/angular.js:7667:11)
   at compositeLinkFn (https://127.0.0.1:8384/vendor/angular/angular.js:7159:13)
   at publicLinkFn (https://127.0.0.1:8384/vendor/angular/angular.js:7038:30)
   at Anonymous function (https://127.0.0.1:8384/vendor/angular/angular.js:1479:11)
   at Scope.prototype.$eval (https://127.0.0.1:8384/vendor/angular/angular.js:14589:9)
```

The problem is about using backticks in https://github.com/syncthing/syncthing/pull/8417/files#diff-77bf455925e2b856359d0a3f63bd07b75fe2990adf70d4996de0df0a0cfe8b29R1546. The fix is at https://stackoverflow.com/a/43048518.

The other error still persists though, although the GUI does load now.

```
SCRIPT1002: Syntax error
webauthn-json.browser-global.js (1,5)
```

**Comment by bt90** [comment]
As we're adding a proper loginpage (yay!) could we also fix the favicon for password managers while we're at it?

https://github.com/syncthing/syncthing/pull/7676

tl;dr provide a static /favicon.ico without requiring authentication

**Comment by emlun** [comment]
>The problem is about using backticks

Ah, of course. Will fix, thanks!

> This is unrelated, but I'm seeing the following in all browsers [...] with the cryptic `%currentDomain%`.

I believe this is because the translation string isn't in place, at least it seemed to fix itself when I added them. You can try running on commit 9dea90adc4a14c339adf8534c5c774389aefa977 to confirm.

> As we're adding a proper loginpage (yay!) could we also fix the favicon for password managers while we're at it?

Hmaybe? The favicon assets `/assets/img/favicon-*.png` are now also in the unauthenticated section, but it looks like we're not actually serving a `/favicon.ico`, instead it's a `<link rel="shortcut icon">` with a dynamic path. But if we just add a static `/favicon.ico`, that will also be available without authentication.

**Comment by bt90** [comment]
At least for Bitwarden it should be enough to have the initial favicon in the HTML without requiring JS and the image to be reachable without authentication.

File: lib/config/guiconfiguration.pb.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (emlun) [reply lib/config/guiconfiguration.pb.go]: Alright, thanks! Should be fixed now.

**Comment by emlun** [comment]
Alright, I've fixed some of the issues encountered so far, and added a todo checklist to the top post. @tomasz1986 please take another look at the support for older browsers! I'll be back with updates for the remaining todo items.

@bt90 I'm pretty sure the favicon can be done quite easily once the auth-split router is in place, but it seems out of scope for this PR.

**Comment by bt90** [comment]
> but it seems out of scope for this PR.

All good :)

**Comment by tomasz1986** [comment]
> @tomasz1986 please take another look at the support for older browsers! I'll be back with updates for the remaining todo items.

I can confirm that now the GUI loads fine in IE11 and displays a message under the authentication that the browser is unsupported. The same is the case with Palemoon, which is another unsupported browser.

However, when actually trying to add new credentials using Chromium, I kept getting the following error in the GUI:

image

and in the browser's console:
```
Credential creation failed. DOMException: This is an invalid domain.
```

This was caused by using `127.0.0.1` as the ID, and the problem disappeared after replacing it with `localhost`. This had to be done in the URL in the browser's address bar too though, however Syncthing by default uses `127.0.0.1`, not `localhost`. Is this expected?

**Comment by emlun** [comment]
> This was caused by using `127.0.0.1` as the ID, and the problem disappeared after replacing it with `localhost`. This had to be done in the URL in the browser's address bar too though, however Syncthing by default uses `127.0.0.1`, not `localhost`. Is this expected?

Ah right, yes, that is expected. Raw IP addresses are not considered "secure contexts", so WebAuthn doesn't work on them. I've added a few interactive warnings about that, and I'll cover it in the documentation too.

File: lib/api/api.go [hunk 2]
```diff
@@ -375,10 +367,24 @@ func (s *service) Serve(ctx context.Context) error {
...
+		mainMux: http.NewServeMux(), // Unauthenticated section
+	}
+	mainMuxBuilder.Handle("/rest/", noCacheRestMux)
+	mainMuxBuilder.HandleFunc("/qr/", s.getQR)
+
+	// Handle the special meta.js path
+	mainMuxBuilder.HandleFunc("/meta.js", s.getJSMetadata)
+
+	// Serve compiled in assets unless an asset directory was set (for development)
+	mainMuxBuilder.mainMux.Handle("/", s.statics)
```
[the remark this answers is not in the document]
Comment (emlun) [reply lib/api/api.go]: In the latest update I've done away with this again. Instead I've moved all (I think) the static resources under the `/static/` prefix which makes it much easier to split out from the authenticated section.

**Comment by emlun** [comment]
Hm, an unexpected side-effect of axing HTTP Basic Auth: now when the server restarts, the session cookie expires in the backend and there's no password cached in the browser, so you get booted back to the login page. And even if we do keep using HTTP Basic Auth behind the scenes, that won't work if you've logged in via WebAuthn.

How best to solve this? Make session cookies survive restart (but time-limited)? Just leave it be because it's a rare enough occurrence?

**Comment by tomasz1986** [comment]
> even if we do keep using HTTP Basic Auth behind the scenes

Just a question related to my personal use case, but I currently rely on being able to log in into Syncthing by embedding the username and password in the URL (e.g. `https://username:password@10.0.0.2:8384`), so that I can have "semi" passwordless local access while still having the GUI protected when accessed remotely. Removing basic authentication means that this will no longer work, right? I've tested the current iteration of the code, and indeed embedding the two in the URL doesn't seem to do anything.

**Comment by emlun** [comment]
>Removing basic authentication means that this will no longer work, right?

Hm. Basic auth was meant to keep working, but it turns out browsers (at least Firefox and Chrome) refuse to send the `Authorization` header if not challenged by a `WWW-Authenticate` response header. Also the `index.html` handler doesn't currently process the `Authorization` header either. But I have gotten this to work again by adding an advanced configuration setting, see: emlun/syncthing#1. Does that work with your use case?

**Comment by tomasz1986** [comment]
> Does that work with your use case?

It works when compiled from the linked PR (and enabling the setting) 👍.

File: gui/default/index.html [hunk 3]
```diff
@@ -330,9 +330,67 @@ <h3 class="panel-title">
         </div>
       </div>
 
-      <!-- First regular row -->
+      <!-- Login form -->
+      <div class="center-block" ng-if="!authenticated">
+        <h3 translate>Authentication required</h1>
```
Comment (tomasz1986) [inline gui/default/index.html:335]:
```suggestion
        <h3 translate>Authentication required</h3>
```

File: gui/default/index.html [hunk 4]
```diff
@@ -330,9 +330,67 @@ <h3 class="panel-title">
...
+        <div class="col-md-6" style="text-align: center">
+          <p translate>
+            Or...
+          </p>
+          <button type="button" class="btn btn-primary" ng-click="authenticateWebauthn()" ng-disabled="login.inProgress || isRawIpAddress() || getLocationProtocol() !== 'https'" tabindex="3">
+            <span translate>Log in with WebAuthn</span>
+          </button>
+
+          <p class="text-info" ng-if="isRawIpAddress()" translate>
+            WebAuthn is not supported on raw IP address hosts.
```
Comment (tomasz1986) [inline gui/default/index.html:372]:
```suggestion
            WebAuthn is not supported on raw IP addresses. If connected locally using <code>127.0.0.1</code>, replace it with <code>localhost</code> instead.
```

Just a suggestion, as the wording can probably be better. The current message is too high-level for a non-techy user to understand what the problem is about.

File: gui/default/index.html [hunk 5]
```diff
@@ -330,9 +330,67 @@ <h3 class="panel-title">
...
+            This device or security key is not registered. Please try a different one.
+          </p>
+          <p class="text-info" ng-if="webauthn.errors.aborted" translate>
+            Authentication aborted by user.
+          </p>
+          <p class="text-danger" ng-if="webauthn.errors.notAllowed" translate>
+            Authentication failed - possibly aborted, timed out, or blocked by the browser.
+          </p>
+          <p class="text-danger" ng-if="webauthn.errors.authenticationFailed" translate>
+            Authentication failed, see server log for details.
```
Comment (tomasz1986) [inline gui/default/index.html:387]: What "server" is being referred here specifically?

File: gui/default/index.html (hunk 4 above)
Comment (emlun) [reply gui/default/index.html:372]: Yes, I've thought about this too, but I've leaned towards covering this in the docs and adding a link to that instead. I hope the new "reload page at WebAuthn Origin" button I added yesterday can help with this as well.

Comment (tomasz1986) [reply gui/default/index.html:372]: The Docs have their limits though, mainly that they're available in English only, while here the strings are translated into all supported languages. If more changes are coming, let's wait for the final version then!

File: gui/default/index.html (hunk 5 above)
Comment (emlun) [reply gui/default/index.html:387]: The Syncthing background service. I've also been thinking this should perhaps make less of a promise that more information will actually be available. How about this?

```suggestion
            Authentication failed. More information may be available in the Syncthing service logs or the browser's developer console.
```

Comment (tomasz1986) [reply gui/default/index.html:387]: How about just "in the Syncthing log"?

File: gui/default/index.html (hunk 4 above)
Comment (emlun) [reply gui/default/index.html:372]: Ah, I see, that's a fair point!

The new button is already pushed in commit https://github.com/syncthing/syncthing/pull/8417/commits/9bc23f51e41b85b7b985c62ab5def326dbe5439b :slightly_smiling_face: Ah, but I see now that this is on the login page, which does not have the button. But I guess we could add one here too - we can even query the back-end for the RP ID (this is already a step in the normal WebAuthn authentication flow, so it won't weaken security) and use that as an even better heuristic for the correct address than just replacing `127.0.0.1` with `localhost`.

**Comment by acolomb** [comment]
Glancing at the diff, I see a rather large risk in moving so much stuff under the `/static/` URL component.  It's very easy to miss code places related to that change.  Is that absolutely necessary, or could we keep the files where they are now?

**Review by acolomb — CHANGES_REQUESTED** [review]
Sorry for taking so long, I just found my review comments still "pending".

**Comment by emlun** [comment]
> Sorry for taking so long

Likewise, sorry for taking so long. I'll try to get back on this soon. :)

> Glancing at the diff, I see a rather large risk in moving so much stuff under the `/static/` URL component. It's very easy to miss code places related to that change. Is that absolutely necessary, or could we keep the files where they are now?

Yeah, I agree with that concern. It's not _absolutely_ necessary, as you can see [an earlier state of the diff as of commit 646b893b](https://github.com/syncthing/syncthing/pull/8417/files/646b893bf554a9785d786b7a75566f77e2f61685#diff-e2e14f22d818b8e635572ef0ee7718dee875c365e07225d760a6faae8be7772d) - look especially at `api.go`. But as you can see there, it takes a pretty complicated tangle of routers to keep the static assets unauthenticated (which is needed since WebAuthn needs an HTML page to be served on) while making sure everything else is behind auth. I judged the risk of breaking some static asset references less significant than the risk of some API paths spilling outside the authenticated API segment - the former would at worst make things not work, and be easy to fix once discovered; the latter would at worst lead to security breaches. But of course I'll revert the change if you prefer keeping the static assets in place.

File: gui/default/assets/css/overrides.css [hunk 6]
```diff
@@ -107,6 +107,10 @@ ul+h5 {
     white-space: nowrap;
 }
 
+.table td.stretch {
+  width: 100%;
```
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/assets/css/overrides.css:111]: Whoops, yes, sorry, I try to do that but clearly I missed it here.

File: gui/default/syncthing/app.js [hunk 7]
```diff
@@ -16,23 +16,26 @@ var syncthing = angular.module('syncthing', [
...
 
 var urlbase = 'rest';
+var authnUrlbase = 'authn';
 
 syncthing.config(function ($httpProvider, $translateProvider, LocaleServiceProvider) {
-    var deviceIDShort = metadata.deviceID.substr(0, 5);
-    $httpProvider.defaults.xsrfHeaderName = 'X-CSRF-Token-' + deviceIDShort;
-    $httpProvider.defaults.xsrfCookieName = 'CSRF-Token-' + deviceIDShort;
-    $httpProvider.useApplyAsync(true);
+    if (window.metadata) {
```
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/syncthing/app.js:22]: I usually prefer lexical control flow over imperative flow, but sure, I can do that.

Hmyeah, I'll need to think a bit on whether some more sophisticated error handling is needed to handle the locale/language settings. I'm not sure we can access those settings before authenticating, or whether that should be allowed. But apart from that it seems appropriate that most everything else remains uninitialized until after authentication.

File: gui/default/syncthing/core/syncthingController.js [hunk 8]
```diff
@@ -13,13 +13,21 @@ angular.module('syncthing.core')
         var restarting = false;
 
         function initController() {
-            LocaleService.autoConfigLocale();
-            setInterval($scope.refresh, 10000);
-            Events.start();
+            if (window.metadata) {
```
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/syncthing/core/syncthingController.js:16]: Will do. The reason here is that the refresh just crashes and spams the error console since we're not authenticated yet. And like above, it might be appropriate to still initialize the locale service.

File: gui/default/syncthing/core/syncthingController.js [hunk 9]
```diff
@@ -83,6 +94,38 @@ angular.module('syncthing.core')
...
+        $scope.getLocationHost = function () {
+            return $location.host();
+        };
+
+        $scope.getLocationProtocol = function() {
+            return $location.protocol();
+        };
+
+        $scope.isRawIpAddress = function (host) {
+            return (host || $location.host()).match(/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}(:.*)?$/) !== null;
```
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/syncthing/core/syncthingController.js:126]: Good catch, hadn't thought of that. Perhaps this should be defined the other way around (checking for a domain address), but I'm not sure there's a simple way to test for that.

**Comment by acolomb** [comment]
This could probably benefit from the `/rest/noauth/` prefix introduced in #8585.

**Comment by bt90** [comment]
@emlun are you still working on this?

**Comment by emlun** [comment]
@bt90 Sorry it's been a while, but I'm looking into this again now. For starters I'll rebase onto the current upstream main and see where to go from there.

**Comment by bt90** [comment]
I'm not sure if it makes sense, but you could split the PR and create one for the login page and another for the webauthn logic.

**Comment by emlun** [comment]
@acolomb Thanks, that did actually help! I could revert most of my changes to the API router stack and just add a few more exceptions to the one added in #8585. I hope this resolves your concerns with moving things under `/static/`; as far as I can tell I managed to revert all those changes too.

I think I've addressed all of the review comments now, you're all welcome to re-review! Tests and documentation still remain to do.

File: lib/api/api.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (emlun) [reply lib/api/api.go]: The latest update reverts most of this and instead adds some more exceptions alongside those added in #8585.

File: gui/default/index.html (hunk 4 above)
Comment (emlun) [reply gui/default/index.html]: I've now added a button to reload the page at (probably) the correct address. There'll be some coverage of this in the docs PR too.

File: gui/default/index.html (hunk 5 above)
Comment (emlun) [reply gui/default/index.html]: Thanks, done now.

File: gui/default/syncthing/core/syncthingController.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/syncthing/core/syncthingController.js]: Done now.

File: gui/default/syncthing/core/syncthingController.js [hunk 10]
```diff
@@ -83,6 +94,38 @@ angular.module('syncthing.core')
...
+            }).finally(function () {
+                $scope.login.inProgress = false;
+            });
+        };
+
+        $scope.getLocationHost = function () {
+            return $location.host();
+        };
+
+        $scope.getLocationProtocol = function() {
```
[the remark this answers is not in the document]
Comment (emlun) [reply gui/default/syncthing/core/syncthingController.js]: Done, thanks!

**Comment by tomasz1986** [comment]
I was going to do a quick test right now, however I'm getting stuck right at the very beginning.

On the first try, I input the username and password, then clicked save, then immediately refreshed the page. The GUI got stuck at the standard login prompt. However, the username and password weren't working at all.

I decided to try again. I wiped out the config and started from scratch. I followed the same steps, however after saving the changes, I waited a few seconds and only then refreshed the GUI. This time the new login screen loaded. I logged in, then opened the settings and removed both username and password, then saved the changes, then refreshed the GUI, and this time again the GUI got stuck at the standard login prompt 😵.

image


image

**Comment by emlun** [comment]
Thanks @tomasz1986! I'll look into these issues tomorrow.

**Comment by tomasz1986** [comment]
Please ignore the above. The problem was that in the past, when launching a new instance of Syncthing, it used to select a random port for the GUI if the default `8384` was in use. However, for some time, the new instance has launched with the same `8384` port which works for its first load in the browser, however if you refresh the site, it tries to log in to the original instance of Syncthing, and thus it fails. Not sure why a random port isn't used anymore, but I don't think the problem is related to this PR at all 🙂.

When it comes to the actual testing, here are some of my comments and questions. I've also added a few others directly in the code.

Firstly, when it comes to the HTML.

1. "Actions -> Settings" doesn't open at all.
2. "Actions -> Show ID" opens but shows nothing.
3. "Actions -> Shutdown" doesn't work at all.
4. "Actions -> Restart" appears to be working but in reality nothing happens.
5. "Actions -> About" opens but the version string and paths are missing.
6. "Actions -> Advanced" doesn't open at all.
7. "Actions -> Logs" opens but shows nothing.
8. "Documentation" link has no version number in it (i.e. `https://docs.syncthing.net/?version=`).
9. Is it unavoidable to use embedded styles with `style` attribute instead of classes with CSS?
10. Is the hard-coded `tabindex` necessary? Which elements get focused first if you remove it?
11. The `translate` attribute is usually as the last in the element. Very minor, but it makes it easier to spot which elements are set to translate and which are not.

Now, the above were just a few comments, however, this is the big one. How does the new login cooperate with the Android app? Not sure if you're aware, but the app currently uses a hard-coded username of "syncthing" and the password set to the API key. The user isn't allowed to change them (or they can try, but the two get reversed upon restart).

1. Does the app still work as before with this PR implemented?
2. If the app itself works, what about the Web GUI?
3. If the Web GUI works, what happens if the user attempts to add new credentials? Is the app still able to revert them, or does doing so break the app (the worst case scenario)?

**Comment by bt90** [comment]
At least from my understanding, only the Basic Auth *prompt* is removed. So the web UI integration should still behave seamlessly:

> Basic authorization headers still work as before

Keeping the credentials locked is valid concern though.

File: lib/config/wrapper.go [hunk 11]
```diff
@@ -92,6 +93,7 @@ type Wrapper interface {
 	RawCopy() Configuration
 	RequiresRestart() bool
 	Save() error
+	Finish(w http.ResponseWriter, waiter Waiter)
```
Comment (imsodin) [inline lib/config/wrapper.go:96]: This is an api specific utility, i.e. shouldn't live in the config package and remain in the api package.  
Incidentlally it's also what makes CI break (and tests not run), as you'd need to run `go run build.go testmocks` to regenerate the mock struct for this interface. However not needed when just not adding it here.

File: lib/api/api_auth.go [hunk 12]
```diff
@@ -210,3 +315,91 @@ func iso88591ToUTF8(s []byte) []byte {
...
+				cfg.GUI.WebauthnCredentials[i].SignCount = updatedCred.Authenticator.SignCount
+				break
+			}
+		}
+	})
+	s.cfg.Finish(w, waiter)
+
+	if updatedCred.Authenticator.CloneWarning && signCountBefore != 0 {
+		l.Warnln(fmt.Sprintf("Invalid WebAuthn signature count for credential \"%s\": expected > %d, was: %d. The credential may have been cloned.", authenticatedCredName, signCountBefore, parsedResponse.Response.AuthenticatorData.Counter))
+	}
```
Comment (imsodin) [inline lib/api/api_auth.go:401]: Can you explain what this counter mechanism does?

Also the messages mentions a comparison to the auth response thing, but the actual condition is just `> 0`.

And the counter isn't config, right? that should go into the db not the config.

File: go.mod [hunk 13]
```diff
@@ -59,25 +59,88 @@ require (
 )
 
 require (
+	github.com/calmh/incontainer v0.0.0-20221224152218-b3e71b103d7a
+	github.com/duo-labs/webauthn v0.0.0-20221205164246-ebaf9b74c6ec
```
Comment (imsodin) [inline go.mod:63]: This is an archived repo, with it's readme point out this as it's successor: https://github.com/go-webauthn/webauthn

**Comment by imsodin** [comment]
I just had a quick look over this PR to get a high level understanding of what's going on, and trying to get into the weeds of the api backend implementation I can only echo this:

> I'm not sure if it makes sense, but you could split the PR and create one for the login page and another for the webauthn logic.

It would help me a lot with review if this were broken up into smaller bits - single PR and multiple clean commits or multiple PRs. From a quick glance splitting like proposed in login page (possible pure refactor first?) and webauthn should work and makes sense semantically - anything else is fine with me too. I know what you really want is webauthn - I promise I'll review both (when I review either - can't promise anything there, wouldn't be honest with my current track record. And you can easily enforce it by not addressing my review if I were to ignore the webauthn bit :) ).

**Comment by emlun** [comment]
@tomasz1986 

> Please ignore the above [...] I don't think the problem is related to this PR at all slightly_smiling_face.

Ok, thanks, glad to hear it. :slightly_smiling_face:

> When it comes to the actual testing, here are some of my comments and questions. [...]

Thanks, I'll look into them!

>Now, the above were just a few comments, however, this is the big one. How does the new login cooperate with the Android app? [...]

Hm, I haven't tested, but I assume the app uses HTTP Basic Authorization then? In that case I think the app should be unaffected, but for it to continue working in a standard browser it probably needs the [advanced setting](https://github.com/syncthing/syncthing/pull/8417#issuecomment-1186236670) to be enabled by default in the server. The new login form should work, of course, but you'd need to know that API key and hard-coded username and without the setting it probably won't work to just open a URL with embedded basic auth from within the app.

**Comment by emlun** [comment]
@bt90 @imsodin Ok, I'll reduce this PR to just adding the new login form and split off the WebAuthn additions to a separate PR. Or would you prefer closing this one and opening two completely new ones?

File: go.mod (hunk 13 above)
Comment (emlun) [reply go.mod:63]: Yeah, I noticed but haven't gotten around to migrating it yet. I'll put it on the todo list.

File: lib/api/api_auth.go (hunk 12 above)
Comment (emlun) [reply lib/api/api_auth.go:401]: It's an optional feature that WebAuthn authenticators can implement, in which case every authentication signature by the credential will include a monotonically increasing counter value. So if a new authentication signature has a counter value less than or equal to the last seen counter value, then that means there likely exists two copies of the credential private key. But the feature is optional, and authenticators that don't implement it instead leave the counter constant at zero, so the check is only made if either the new or the last seen counter value is nonzero. You can read more about it [here](https://w3c.github.io/webauthn/#signature-counter). But the library [encapsulates all that in the `CloneWarning` output](https://github.com/duo-labs/webauthn/blob/092d5e0bbb16390d0cb46851c433218ac4a45d7a/webauthn/authenticator.go#L46-L47), so the zero check here is unrelated.

I think the `!= 0` check here is just because as far as I understand, the Go convention for searching in a list is to initialize to a sentinel value and use a for loop (unlike, say, Rust where you'd `.find(...).map(...) -> Option<u32>`). I'd love to be wrong about that and learn there's a better way, though. :slightly_smiling_face:

> And the counter isn't config, right? that should go into the db not the config.

Hmyeah, I guess so? In that case the whole credential entries should perhaps go in the database instead, as it's not really feasible to type them in manually either. How do you reckon that would work with configuring them in the settings view? Where should I start looking at how to work with the database?

File: lib/config/wrapper.go (hunk 11 above)
Comment (emlun) [reply lib/config/wrapper.go:96]: Ok, thanks, I'll see if I can refactor this.

**Comment by imsodin** [comment]
Whatever works best for you - reducing this one is perfectly fine for me

**Comment by tomasz1986** [comment]
> Hm, I haven't tested, but I assume the app uses HTTP Basic Authorization then? In that case I think the app should be unaffected, but for it to continue working in a standard browser it probably needs the [advanced setting](https://github.com/syncthing/syncthing/pull/8417#issuecomment-1186236670) to be enabled by default in the server. The new login form should work, of course, but you'd need to know that API key and hard-coded username and without the setting it probably won't work to just open a URL with embedded basic auth from within the app.

Yeah, I don't know the details when it comes to the app code, but if the advanced setting is indeed required, then a) it would need to be enabled automatically upon upgrade or otherwise the user may be stuck with inaccessible Syncthing, and b) the setting would have to be enforced, so that the user basically can't change it (similarly to the currently hard-coded username and password). Basically, the app mustn't allow the user to break it 😛. This is especially true as once locked out, the user probably won't know the API key, and in contrary to the desktop, there is no way to even access `config.xml` on Android to restore it (without root).

**Comment by acolomb** [comment]
> This is especially true as once locked out, the user probably won't know the API key, and in contrary to the desktop, there is no way to even access `config.xml` on Android to restore it (without root).

Well I suppose the app itself does have access to the file, so could of course restore access when the service is not running.

**Comment by emlun** [comment]
I realized I can't change the branch name of this PR, and it doesn't make much sense to have branch `webauthn` contain nothing related to WebAuthn :) so I extracted a new branch and PR with only the new login page: #8757. Should I also make a concurrent PR (either this or a new one) that isolates the changes for adding WebAuthn?

**Comment by imsodin** [comment]
> I realized I can't change the branch name of this PR, and it doesn't make much sense to have branch `webauthn` contain nothing related to WebAuthn :) so I extracted a new branch and PR with only the new login page: #8757. Should I also make a concurrent PR (either this or a new one) that isolates the changes for adding WebAuthn?

Whatever works best for you. If the webauthn PR builds on top of the login page one, you might want to PR against the login page branch. Then only the relevant changes are shown in the PR, and once the login page one is merged it will automatically target main.

**Comment by st-review** [comment]
:robot: beep boop

I'm going to close this pull request as it has been idle for more than 90 days.

This is not a rejection, merely a removal from the list of active pull requests that is periodically reviewed by humans. The pull request can be reopened when there is new activity, and merged once any remaining issues are resolved.

To permanently exempt this PR from bot nagging, add the `slow-pr` label.

**Comment by emlun** [comment]
#8757 is now merged! With that I'll (soon :sweat_smile:) proceed to revive this with the remaining additions needed to also support WebAuthn. I'll probably start fresh with a new pull request, but I'll post a link here for the record.

**Comment by emlun** [comment]
This PR is now replaced by PR #9175.
